File: assignment-1/solutions/task1.py
import json
import sys
from datetime import datetime

from pyspark import SparkContext


def task1(review_file_path, output_file_path):
    spark = SparkContext(appName="Yelp Data Exploration (Task 1)").getOrCreate()

    review_rdd = spark.textFile(review_file_path).map(json.loads).cache()

    result = {
        "n_review": 0,
        "n_review_2018": 0,
        "n_user": 0,
        "top10_user": [],
        "n_business": 0,
        "top10_business": [],
    }

    try:
        result["n_review"] = review_rdd.count()

        result["n_review_2018"] = review_rdd.filter(
            lambda x: datetime.strptime(x["date"], "%Y-%m-%d %H:%M:%S").year == 2018
        ).count()

        result["n_user"] = review_rdd.map(lambda x: x["user_id"]).distinct().count()

        result["top10_user"] = (
            review_rdd.map(lambda x: (x["user_id"], 1))
            .reduceByKey(lambda a, b: a + b)
            .sortBy(lambda x: (-x[1], x[0]), ascending=True)
            .take(10)
        )

        result["n_business"] = review_rdd.map(lambda x: x["business_id"]).distinct().count()

        result["top10_business"] = (
            review_rdd.map(lambda x: (x["business_id"], 1))
            .reduceByKey(lambda a, b: a + b)
            .sortBy(lambda x: (-x[1], x[0]), ascending=True)
            .take(10)
        )

        with open(output_file_path, "w") as f:
            json.dump(result, f, indent=4)

    finally:
        spark.stop()


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print(
            "Usage: spark-submit --executor-memory 4G --driver-memory 4G "
            "task1.py <review_file_path> <output_file_path>"
            " --conf spark.driver.extraJavaOptions=-Dlog4j.configuration=log4j.properties"
        )
        sys.exit(1)

    review_file_path = sys.argv[1]
    output_file_path = sys.argv[2]
    task1(review_file_path, output_file_path)

File: assignment-1/solutions/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._

object task1 {
  def main(args: Array[String]): Unit = {
    if (args.length != 2) {
      println(
        "Usage: spark-submit --executor-memory 4G --driver-memory 4G task1.py <review_file_path> <output_file_path>"
      )
      sys.exit(1)
    }

    val reviewFilePath = args(0)
    val outputFilePath = args(1)

    val spark = SparkSession
      .builder()
      .appName("Yelp Data Exploration (Task 1)")
      .getOrCreate()

    try {
      val reviewRDD = spark.sparkContext
        .textFile(reviewFilePath)
        .map(row => row.split("\\|")) // Assuming the data is pipe-separated

      val nReview = reviewRDD.count()

      val nReview2018 = reviewRDD
        .filter(row => {
          val date = row(3)
          val year = date.split(" ")(0).split("-")(0).toInt
          year == 2018
        })
        .count()

      val nUser = reviewRDD.map(row => row(1)).distinct().count()

      val top10User = reviewRDD
        .map(row => (row(1), 1))
        .reduceByKey(_ + _)
        .sortBy(_._2, ascending = false)
        .take(10)

      val nBusiness = reviewRDD.map(row => row(2)).distinct().count()

      val top10Business = reviewRDD
        .map(row => (row(2), 1))
        .reduceByKey(_ + _)
        .sortBy(_._2, ascending = false)
        .take(10)

      val result = Map(
        "n_review" -> nReview,
        "n_review_2018" -> nReview2018,
        "n_user" -> nUser,
        "top10_user" -> top10User,
        "n_business" -> nBusiness,
        "top10_business" -> top10Business
      )

      import spark.implicits._
      val resultDF = result.toList.toDF("Metric", "Value")

      resultDF.write
        .format("json")
        .mode("overwrite")
        .save(outputFilePath)
    } finally {
      spark.stop()
    }
  }
}
